active/0000-lazy-load-overview.md

```diff
@@ -0,0 +1,185 @@
+and components. It is also important to consider asset dependencies.
+
+## Building
+
+We need to change the way we build our assets. By default Ember CLI bundles everything in two app and vendor assets for
+CSS and JS. In order to be able to lazy-load, we need to allow for customizations to how we build. This is partially
+possible today, but further enhancements are required.
+
+### Building multiple CSS:
+
+Today we have two ways of building multiple CSS assets. For vendor static assets (from `vendor/` or `bower_components`)
+we can import them to a different `outputFile` as specified on
+[RFC#28](https://github.com/ember-cli/rfcs/pull/28) and supported since ember-cli 2.4.
+
+```
+app.import(app.bowerDirectory + '/bootstrap/dist/css/bootstrap.css', {outputFile: 'vendor2.css'});
+```
+
+For files that need a pre-processor we can use `outputPaths`. In the example below, `styles/deferred-styles.scss` is
+simply a file with references to other CSS or SASS files. This works for `vendor` and `app` styles.
+
+```
+let app = new EmberApp(defaults, {
+  outputPaths: {
+    app: {
+      css: {
+        'app': '/assets/css/app.css',
+        'deferred-styles': '/assets/css/deferred-styles.css',
+      }
+    }
+  }
+});
+```
+
+### Building multiple vendor JS assets from static files (e.g. bower/vendor):
+
+For vendor static assets (from `vendor/` or `bower_components`) we can import them to a different `outputFile`
+as specified on [RFC#28](https://github.com/ember-cli/rfcs/pull/28) and supported since ember-cli 2.4.
+
+
+```
+app.import('vendor/dependency-1.js', { outputFile: 'assets/alternate-vendor.js'});
+```
+
+### Building multiple vendor JS assets from addons
+
+There is an [RFC](https://github.com/ember-cli/rfcs/pull/52) for this, but the concept is similar to the above, but
+extended for addons.
+
+### Building multiple JS assets for app code
+
+This isn't supported out of the box today, but it's possible to do. There are two example,
+[ember-cli-bundle-loader](https://github.com/MiguelMadero/ember-cli-bundle-loader) and
+[ember-lazy-loader](https://github.com/duizendnegen/ember-cli-lazy-load). The first one relies on a different file structure 
+and uses multiple ember-apps to build each asset. The latter relies on configuration to specify which app files will go to each
+asset.
+
+This RFC is limited to provide an overview of what is required, so this topic requires a bit more discusion on a
+separate RFC. TODO: create the separate RFC. Engines can provide a nice default, where we can have a convention of
+one asset per engine when lazyLoading is enabled, which can be overriden to allow for one or more engines per file.
+
+## Loading assets
+
+[ember-cli-bundle-loader](https://github.com/MiguelMadero/ember-cli-bundle-loader) provides a
+[service](https://github.com/MiguelMadero/ember-cli-bundle-loader/blob/master/addon/services/lazy-loader.js)
```

Comment:
I'll look into it. 

>  it requires tight coupling to the build-pipeline 

I think it makes more sense to decouple them, but let me think it through a bit more. There are many ways in which we can build assets today from CLI, engines will add one more and we have cases where we might get assets from CDNs or generated by a different asset-pipeline. I think we need a manifest (for all CLI assets, similar to what we get from AssetRev) an extension point, so users can extend it for external assets, but the service can focus on loading JS/CSS in a reliable promise-aware way that can be used by components and the router. 

- [ ] Extract AssetService from bundle-loader and write RFC.